Add sample .bats files into the examples/ dir

amilkh opened the pull request:
This will help beginners integrate bats into their workflow more easily.

Authored-by: Amil Khanzada <akhanzada@pivotal.io>

- [x] I have reviewed the [Contributor Guidelines][contributor].
- [x] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

alexdesignworks commented on examples/if_ls_glob.bats:
```diff
@@ -0,0 +1,49 @@
+}
+
+@test "it works when there are three .bats file" {
+  touch {a,b,c}.bats
```
is it possible to not use `.bats` extension as for the beginners it is not clear why `.bats` was used specifically. Using `.txt` would be more straightforward.

alexdesignworks commented on examples/if_ls_glob.bats:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bats
+
```
could you please add a description that this is an example file

amilkh replied:
hmm... I think the file should be self-documenting after the latest refactors. I try to avoid comments when the code can self-document.

alexdesignworks commented on examples/if_ls_glob.bats:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bats
+
+TEST_DIR="test_dir"
```
it is not clear what this variable for.

here and below - can you please add comments for every part that is not documented, so that newcomers could read the file without guessing what every line means, thanks

amilkh replied:
thanks commend added!

alexdesignworks commented on examples/if_ls_glob.bats:
```diff
@@ -0,0 +1,49 @@
+
+main() {
+  if ls *.bats >/dev/null 2>&1; then
+    bats -t *.bats
```
could you please explain why does the test function need to run tests. this is inception. very confusing.

can this simply create a set of fixture files?

or maybe I'm missing the point about this test example (hence the file comment would help to set the context)

amilkh replied:
have removed the `ls` script for now because it is confusing!

amilkh commented:
Thanks for the feedback @alexdesignworks ! I'll address your comments on Friday.

amilkh commented:
Have made some updates, please let me know your thoughts!

alexskrypnyk commented:
@amilkh re p.3:
1. Latest version of shellcheck supports `.bats` linting
2. There is a PR for the CI of this project to check all files.

amilkh commented:
> 1. Latest version of shellcheck supports `.bats` linting
> 2. There is a PR for the CI of this project to check all files.

Thanks! I ran shellcheck.
also added note to README.md


if this PR looks ok, can we get it merged?

virgilwashere commented:
> Questions:
> 
> 1. From the code guidelines it says:
>> Declare temporary file-level variables using declare. Use unset to remove them when finished.
> should this apply to `$TEST_DIR`?

Yes it should. However ...(see next)

> 2. I think there may be a better pattern than `$TEST_DIR`.

Use `$BATS_TMPDIR` as defined in [Special Variables](/bats-core/bats-core/blob/master/README.md#special-variables)

- [x] pre-declared for you

- [x] existing standard

And I would suggest using `$BATS_TEST_DIRNAME` in your `main()` function. 


```bash
main() {
  "${BATS_TEST_DIRNAME}/package_tarball" "$@"
}
```

virgilwashere commented on README.md:
```diff
@@ -264,6 +264,8 @@ process.
 For more details about how Bats evaluates test files, see [Bats Evaluation
 Process][bats-eval] on the wiki.
 
+For sample test files, see [examples](/examples).
```
The directory `/docs/examples` seems like a better home for this content.

```suggestion
For sample test files, see [examples](/docs/examples).
```

Include `examples/README.md` which at the very least should contain a link to the [Wiki Examples page](https://github.com/bats-core/bats-core/wiki/Examples).

```markdown
# Examples

See the [bats-core wiki][examples] for more details

[examples]: (/bats-core/bats-core/wiki/Examples)
```

virgilwashere commented on examples/package-tarball:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+set -eux
```
If it's going to be an example, let's follow good principals.

```suggestion
# "unofficial" bash strict mode
# See: http://redsymbol.net/articles/unofficial-bash-strict-mode
set -o errexit  # Exit when simple command fails               'set -e'
set -o errtrace # Exit on error inside any functions or subshells.
set -o nounset  # Trigger error when expanding unset variables 'set -u'
set -o pipefail # Do not hide errors within pipes              'set -o pipefail'
set -o xtrace   # Display expanded command and arguments       'set -x'
IFS=$'\n\t'
```

amilkh commented:
Thanks so much for the detailed review @virgilwashere !!
I've addressed all your comments, I think, so please let me know any other feedback or go ahead an merge!

AlexSkrypnyk commented:
@amilkh thank you for working on this. Looks great!

One of the CI jobs has a false positive fail - need maintainers to restart it manually.

amilkh commented:
Thanks guys for supporting me on this!! Have rebased everything into one commit, so please go ahead and merge when you're ready.

sublimino commented:
Hello @amilkh (and @virgilwashere and @AlexSkrypnyk!) good teamwork on this one : ] Just FYI -- I'll [add this to the next release](https://github.com/bats-core/bats-core/projects/4) and ensure it's merged after the next refactors are done. 

Thanks all for your hard work, this is a nice clean PR with lots of great conversation. Thank you! 🎉
